lib/rubocop/cop/rspec/instance_spy.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for `instance_double(Foo).as_null_object usage`
+      # when the double is used with `have_received`.
+      #
+      # @example
+      # # bad
+      # it do
+      #  foo = instance_double(Foo).as_null_object
+      #  expect(foo).to have_received(:bar)
+      # end
+      #
+      # # good
+      # it do
+      #   foo = instance_spy(Foo)
+      #   expect(foo).to have_received(:bar)
+      # end
+      #
+      class InstanceSpy < Cop
+        MSG = 'Use `instance_spy` when you check your double ' \
+              'with `have_received`'.freeze
+
+        EXAMPLE_BLOCKS = RuboCop::RSpec::Language::Examples::ALL
+
+        def_node_search :spy, <<-PATTERN
+        (lvasgn $_
+          (send
+            $(send nil :instance_double
+              (const nil _)) :as_null_object))
+        PATTERN
+
+        def_node_search :expectation, <<-PATTERN
+        (send
+          (send nil :expect
+          (lvar $_)) :to
+          (send nil :have_received
+          ...)
+        ...)
+        PATTERN
+
+        def on_block(node)
+          method, _args, _body = *node
+          _receiver, method_name, _object = *method
+          return unless EXAMPLE_BLOCKS.include?(method_name)
```

Comment:
You might be able to use `SelectorSet#block_pattern`. Grep for examples to see if it would fit here